sycl/test-e2e/Matrix/joint_matrix_gemm_cuda.hpp

```diff
@@ -30,37 +30,47 @@ constexpr float bf16_eps = 0.00390625;
 constexpr int N_THREADS_PER_MATRIX_OP = 32;
 
 // number of submatrices per row of accumulator ("C", "D") matrices.
-constexpr int SUB_TILES_M = 3;
+constexpr int SUB_TILES_M = 2;
 // number of submatrices per col of accumulator matrices.
 constexpr int SUB_TILES_N = 2;
 // number of submatrices per col of "A"/per row of "B", matrices.
-constexpr int SUB_TILES_K = 1;
+constexpr int SUB_TILES_K = 2;
 
-template <typename Tm, typename Tc, typename Td, size_t M, size_t K, size_t N>
+template <typename Tm, typename Tc, typename Td, size_t M, size_t K, size_t N,
+          layout layout_A, layout layout_B, layout layout_C>
 class TypeHelper;
 
-template <typename Tm, typename Tc, typename Td, size_t M, size_t K, size_t N>
-using KernelName = class TypeHelper<Tm, Tc, Td, M, K, N>;
+template <typename Tm, typename Tc, typename Td, size_t M, size_t K, size_t N,
+          layout layout_A, layout layout_B, layout layout_C>
+using KernelName =
+    class TypeHelper<Tm, Tc, Td, M, K, N, layout_A, layout_B, layout_C>;
 
-template <size_t Big_N, size_t Big_K, typename Tm, typename Tc>
+template <size_t Big_N, size_t Big_K, size_t Big_M, layout layout_A,
+          layout layout_B, typename Tm, typename Tc>
 Tc matrix_ref_mn(const int &m, const int &n, Tm *A, Tm *B, Tc *C) {
```

Comment:
I believe, this function is solving problem similar to what is done here: https://github.com/intel/llvm/blob/48ec4dd6ff45b25b405b99d63f3c5b8537b1475d/sycl/test-e2e/Matrix/common.hpp#L39
Would be good to try to unify both functions, as it seems, they are supposed to be used for the same.

Reply:
I agree, and I was actually thinking of renaming as:

joint_matrix_gemm_cuda.hpp -> joint_matrix_gemm.hpp

to encourage this.

This `test` function is I think the only example in this directory that is completely backend agnostic, and provides all template parameters required to completely test all possible parameters of the core gemm APIs that are supported by all backends that support the oneapi matrix extension; in particular:

- joint_matrix_load
- joint_matrix_store
- joint_matrix_mad

Other backends could use the `test` function in joint_matrix_gemm_cuda.hpp in order to have full coverage testing of their supported types/matrix sizes, exactly as is done in joint_matrix_tensorcore_*.cpp examples.  This would greatly simplify the test directory, and aid maintenance.
The only variable in the header that could need to be changed for different backends is

```
constexpr int N_THREADS_PER_MATRIX_OP = 32;
```

Reply:
if `joint_matrix_gemm_cuda.hpp` doesn't have any cuda specific, I would also vote for renaming `joint_matrix_gemm_cuda.hpp -> joint_matrix_gemm.hpp`

> Then other backends will automatically use the reference function in this header also

Not sure how it will happen automatically. I believe, manual work will be needed to move all existing tests for other backends to new framework. But again, I like the idea of getting rid of duplicated code and using the same framework across backends if practical.

> constexpr int N_THREADS_PER_MATRIX_OP = 32;

This might need to be removed if we go with joint_matrix_gemm.hpp, so this file would not contain any backend specific.
Also since Intel HW supports different sub group sizes, we set attribute [[intel::reqd_sub_group_size(SG_SZ)]] for gemm kernel to test with different sub group sizes, where SG_SZ can be 8, 16, 32 depending on HW. Also it is possible not to set it at all to test with subgroup size selected by compiler. I think this use case would need to be introduced to be able to apply joint_matrix_gemm.hpp for Intel backend tests.

Anyway, cleaning up the file from any CUDA specifics and renaming to `joint_matrix_gemm.hpp` might be a good first step for future unification.

@dkhaldi , what do you think about all that?

Reply:
> 
> > constexpr int N_THREADS_PER_MATRIX_OP = 32;
> 
> This might need to be removed if we go with joint_matrix_gemm.hpp, so this file would not contain any backend specific. Also since Intel HW supports different sub group sizes, we set attribute [[intel::reqd_sub_group_size(SG_SZ)]] for gemm kernel to test with different sub group sizes, where SG_SZ can be 8, 16, 32 depending on HW. Also it is possible not to set it at all to test with subgroup size selected by compiler. I think this use case would need to be introduced to be able to apply joint_matrix_gemm.hpp for Intel backend tests.
> 

HIP hardware that supports joint_matrix also has a different subgroup size to the cuda backend: HIP only supports subgroup size 64 for joint_matrix. I'm happy to unify the HIP testing with this CUDA testing as an example. I think that the intel unification would follow the same process.